dds/src/subscription/data_reader.rs

```diff
@@ -230,13 +268,17 @@ where
         view_states: &[ViewStateKind],
         instance_states: &[InstanceStateKind],
     ) -> DdsResult<Vec<Sample<Foo>>> {
-        self.0.upgrade()?.take(
-            max_samples,
-            sample_states,
-            view_states,
-            instance_states,
-            Some(a_handle),
-        )
+        match &self.0 {
+            DataReaderKind::BuiltinStateless(_) => todo!(),

```

Comment:
That enum seems to make the reader quite complicated wit that many todo()'s

Reply:
It is a bit unfortunate but it is the price to pay to be able to expose the different types under a common interface.